File: internal/scheduler/validate.go
package scheduler

import (
	"fmt"

	"github.com/pseudomuto/skedge/internal/config"
)

// maxTeacherBlocksPerWeek is the hard cap on teaching blocks per teacher across
// all cohorts in a week.
const maxTeacherBlocksPerWeek = 21

// Validate checks all schedule constraints and returns the first violation
// found, or nil if the schedule is valid. The checks are:
//
//   - [ValidateNoTeacherConflict]: no teacher in two places at once.
//   - [ValidateNoRoomConflict]: no room used by two cohorts at once.
//   - [ValidateTeacherBlockLimit]: no teacher exceeds the weekly block cap.
//   - [ValidateNoDuplicateSubjectPerDay]: each subject appears at most once per day per cohort.
//   - [ValidateTeacherSubjects]: teachers only teach subjects they are authorized for.
//   - [ValidateSubjectBlockCounts]: each cohort has the exact weekly block count per subject.
//   - [ValidateNoGaps]: every block slot has a subject, teacher, and room assigned.
func Validate(cfg *config.Config, schedule []*Class) error {
	checks := []func() error{
		func() error { return ValidateNoTeacherConflict(schedule) },
		func() error { return ValidateNoRoomConflict(schedule) },
		func() error { return ValidateTeacherBlockLimit(schedule) },
		func() error { return ValidateNoDuplicateSubjectPerDay(schedule) },
		func() error { return ValidateTeacherSubjects(cfg, schedule) },
		func() error { return ValidateSubjectBlockCounts(cfg, schedule) },
		func() error { return ValidateNoGaps(cfg, schedule) },
	}
	for _, check := range checks {
		if err := check(); err != nil {
			return err
		}
	}
	return nil
}

// ValidateNoTeacherConflict returns an error if any teacher is assigned to
// more than one cohort in the same (day, block) slot.
func ValidateNoTeacherConflict(schedule []*Class) error {
	type key struct {
		day     string
		block   int
		teacher string
	}

	seen := make(map[key]string) // key -> first cohort name
	for _, class := range schedule {
		for _, cohort := range class.Cohorts {
			for _, d := range cohort.Schedule {
				for i, b := range d.Blocks {
					k := key{d.Day, i, b.Teacher}
					if other, ok := seen[k]; ok {
						return fmt.Errorf(
							"teacher %s double-booked on %s block %d: cohorts %s and %s",
							b.Teacher,
							d.Day,
							i,
							other,
							cohort.Name,
						)
					}

					seen[k] = cohort.Name
				}
			}
		}
	}
	return nil
}

// ValidateNoRoomConflict returns an error if any room is used by more than one
// cohort in the same (day, block) slot.
func ValidateNoRoomConflict(schedule []*Class) error {
	type key struct {
		day   string
		block int
		room  string
	}
	seen := make(map[key]string) // key -> first cohort name
	for _, class := range schedule {
		for _, cohort := range class.Cohorts {
			for _, d := range cohort.Schedule {
				for i, b := range d.Blocks {
					k := key{d.Day, i, b.Room}
					if other, ok := seen[k]; ok {
						return fmt.Errorf("room %s double-booked on %s block %d: cohorts %s and %s", b.Room, d.Day, i, other, cohort.Name)
					}
					seen[k] = cohort.Name
				}
			}
		}
	}
	return nil
}

// ValidateTeacherBlockLimit returns an error if any teacher is assigned more
// than [maxTeacherBlocksPerWeek] blocks in total across all cohorts.
func ValidateTeacherBlockLimit(schedule []*Class) error {
	counts := make(map[string]int)
	for _, class := range schedule {
		for _, cohort := range class.Cohorts {
			for _, d := range cohort.Schedule {
				for _, b := range d.Blocks {
					counts[b.Teacher]++
				}
			}
		}
	}
	for teacher, count := range counts {
		if count > maxTeacherBlocksPerWeek {
			return fmt.Errorf("teacher %s has %d blocks, exceeds weekly limit of %d", teacher, count, maxTeacherBlocksPerWeek)
		}
	}
	return nil
}

// ValidateNoDuplicateSubjectPerDay returns an error if any cohort has the same
// subject scheduled more than once on the same day.
func ValidateNoDuplicateSubjectPerDay(schedule []*Class) error {
	for _, class := range schedule {
		for _, cohort := range class.Cohorts {
			if err := validateCohortNoDuplicateSubjectPerDay(&cohort); err != nil {
				return err
			}
		}
	}
	return nil
}

// ValidateTeacherSubjects returns an error if any teacher is assigned a subject
// they are not authorized to teach in the relevant class.
func ValidateTeacherSubjects(cfg *config.Config, schedule []*Class) error {
	type authKey struct{ teacher, class string }
	authorized := make(map[authKey]map[string]struct{})
	for _, t := range cfg.Teachers {
		for _, ts := range t.Subjects {
			k := authKey{t.Name, ts.Class}
			if authorized[k] == nil {
				authorized[k] = make(map[string]struct{})
			}
			for _, s := range ts.Subjects {
				authorized[k][s] = struct{}{}
			}
		}
	}
	for _, class := range schedule {
		for _, cohort := range class.Cohorts {
			for _, d := range cohort.Schedule {
				for _, b := range d.Blocks {
					k := authKey{b.Teacher, class.Name}
					subjects, ok := authorized[k]
					if !ok {
						return fmt.Errorf("teacher %s has no subjects authorized for class %s", b.Teacher, class.Name)
					}
					if _, ok := subjects[b.Name]; !ok {
						return fmt.Errorf("teacher %s not authorized to teach %s in class %s", b.Teacher, b.Name, class.Name)
					}
				}
			}
		}
	}
	return nil
}

// ValidateSubjectBlockCounts returns an error if any cohort's actual weekly
// block count for a subject differs from the count required by the class
// configuration.
func ValidateSubjectBlockCounts(cfg *config.Config, schedule []*Class) error {
	required := make(map[string]map[string]int) // className -> subjectName -> required count
	for _, c := range cfg.Classes {
		required[c.Name] = make(map[string]int)
		for _, s := range c.Subjects {
			required[c.Name][s.Name] = s.Blocks
		}
	}
	for _, class := range schedule {
		req := required[class.Name]
		for _, cohort := range class.Cohorts {
			actual := make(map[string]int)
			for _, d := range cohort.Schedule {
				for _, b := range d.Blocks {
					actual[b.Name]++
				}
			}
			for subject, want := range req {
				if actual[subject] != want {
					return fmt.Errorf("cohort %s has %d blocks of %s, expected %d", cohort.Name, actual[subject], subject, want)
				}
			}
			for subject := range actual {
				if _, ok := req[subject]; !ok {
					return fmt.Errorf("cohort %s has unknown subject %s", cohort.Name, subject)
				}
			}
		}
	}
	return nil
}

// ValidateNoGaps returns an error if any block slot has an empty subject,
// teacher, or room, or if a cohort's day has a different number of blocks than
// the configuration specifies.
func ValidateNoGaps(cfg *config.Config, schedule []*Class) error {
	expected := len(cfg.Blocks)
	for _, class := range schedule {
		for _, cohort := range class.Cohorts {
			for _, d := range cohort.Schedule {
				if len(d.Blocks) != expected {
					return fmt.Errorf("cohort %s on %s has %d blocks, expected %d", cohort.Name, d.Day, len(d.Blocks), expected)
				}
				for i, b := range d.Blocks {
					if b.Name == "" || b.Teacher == "" || b.Room == "" {
						return fmt.Errorf("cohort %s on %s block %d has empty field", cohort.Name, d.Day, i)
					}
				}
			}
		}
	}
	return nil
}

func validateCohortNoDuplicateSubjectPerDay(cohort *Cohort) error {
	for _, d := range cohort.Schedule {
		seen := make(map[string]struct{})
		for _, b := range d.Blocks {
			if _, ok := seen[b.Name]; ok {
				return fmt.Errorf("cohort %s has subject %s more than once on %s", cohort.Name, b.Name, d.Day)
			}
			seen[b.Name] = struct{}{}
		}
	}
	return nil
}

File: web/src/scheduler/validate.ts
import type { Config } from '../types/config'
import type { ScheduleClass } from '../types/schedule'

export const MAX_TEACHER_BLOCKS_PER_WEEK = 21

export function validateNoTeacherConflict(schedule: ScheduleClass[]): Error | null {
  const seen = new Map<string, string>()
  for (const cls of schedule) {
    for (const cohort of cls.cohorts) {
      for (const d of cohort.schedule) {
        for (let i = 0; i < d.blocks.length; i++) {
          const b = d.blocks[i]
          const key = `${d.day}:${i}:${b.teacher}`
          const other = seen.get(key)
          if (other !== undefined) {
            return new Error(
              `teacher ${b.teacher} double-booked on ${d.day} block ${i}: cohorts ${other} and ${cohort.name}`,
            )
          }
          seen.set(key, cohort.name)
        }
      }
    }
  }
  return null
}

export function validateNoRoomConflict(schedule: ScheduleClass[]): Error | null {
  const seen = new Map<string, string>()
  for (const cls of schedule) {
    for (const cohort of cls.cohorts) {
      for (const d of cohort.schedule) {
        for (let i = 0; i < d.blocks.length; i++) {
          const b = d.blocks[i]
          const key = `${d.day}:${i}:${b.room}`
          const other = seen.get(key)
          if (other !== undefined) {
            return new Error(
              `room ${b.room} double-booked on ${d.day} block ${i}: cohorts ${other} and ${cohort.name}`,
            )
          }
          seen.set(key, cohort.name)
        }
      }
    }
  }
  return null
}

export function validateTeacherBlockLimit(schedule: ScheduleClass[]): Error | null {
  const counts = new Map<string, number>()
  for (const cls of schedule) {
    for (const cohort of cls.cohorts) {
      for (const d of cohort.schedule) {
        for (const b of d.blocks) {
          counts.set(b.teacher, (counts.get(b.teacher) ?? 0) + 1)
        }
      }
    }
  }
  for (const [teacher, count] of counts) {
    if (count > MAX_TEACHER_BLOCKS_PER_WEEK) {
      return new Error(
        `teacher ${teacher} has ${count} blocks, exceeds weekly limit of ${MAX_TEACHER_BLOCKS_PER_WEEK}`,
      )
    }
  }
  return null
}

export function validateNoDuplicateSubjectPerDay(schedule: ScheduleClass[]): Error | null {
  for (const cls of schedule) {
    for (const cohort of cls.cohorts) {
      for (const d of cohort.schedule) {
        const seen = new Set<string>()
        for (const b of d.blocks) {
          if (seen.has(b.name)) {
            return new Error(
              `cohort ${cohort.name} has subject ${b.name} more than once on ${d.day}`,
            )
          }
          seen.add(b.name)
        }
      }
    }
  }
  return null
}

export function validateTeacherSubjects(cfg: Config, schedule: ScheduleClass[]): Error | null {
  const authorized = new Map<string, Set<string>>()
  for (const t of cfg.teachers) {
    for (const ts of t.subjects) {
      const k = `${t.name}:${ts.class}`
      if (!authorized.has(k)) authorized.set(k, new Set())
      for (const s of ts.subjects) authorized.get(k)!.add(s)
    }
  }
  for (const cls of schedule) {
    for (const cohort of cls.cohorts) {
      for (const d of cohort.schedule) {
        for (const b of d.blocks) {
          const k = `${b.teacher}:${cls.name}`
          const subjects = authorized.get(k)
          if (!subjects) {
            return new Error(
              `teacher ${b.teacher} has no subjects authorized for class ${cls.name}`,
            )
          }
          if (!subjects.has(b.name)) {
            return new Error(
              `teacher ${b.teacher} not authorized to teach ${b.name} in class ${cls.name}`,
            )
          }
        }
      }
    }
  }
  return null
}

export function validateSubjectBlockCounts(cfg: Config, schedule: ScheduleClass[]): Error | null {
  const required = new Map<string, Map<string, number>>()
  for (const c of cfg.classes) {
    const m = new Map<string, number>()
    for (const s of c.subjects) m.set(s.name, s.blocks)
    required.set(c.name, m)
  }
  for (const cls of schedule) {
    const req = required.get(cls.name)
    if (!req) continue
    for (const cohort of cls.cohorts) {
      const actual = new Map<string, number>()
      for (const d of cohort.schedule) {
        for (const b of d.blocks) {
          actual.set(b.name, (actual.get(b.name) ?? 0) + 1)
        }
      }
      for (const [subject, want] of req) {
        const got = actual.get(subject) ?? 0
        if (got !== want) {
          return new Error(
            `cohort ${cohort.name} has ${got} blocks of ${subject}, expected ${want}`,
          )
        }
      }
      for (const [subject] of actual) {
        if (!req.has(subject)) {
          return new Error(`cohort ${cohort.name} has unknown subject ${subject}`)
        }
      }
    }
  }
  return null
}

export function validateNoGaps(cfg: Config, schedule: ScheduleClass[]): Error | null {
  const expected = cfg.blocks.length
  for (const cls of schedule) {
    for (const cohort of cls.cohorts) {
      for (const d of cohort.schedule) {
        if (d.blocks.length !== expected) {
          return new Error(
            `cohort ${cohort.name} on ${d.day} has ${d.blocks.length} blocks, expected ${expected}`,
          )
        }
        for (let i = 0; i < d.blocks.length; i++) {
          const b = d.blocks[i]
          if (!b.name || !b.teacher || !b.room) {
            return new Error(`cohort ${cohort.name} on ${d.day} block ${i} has empty field`)
          }
        }
      }
    }
  }
  return null
}

export function validate(cfg: Config, schedule: ScheduleClass[]): Error | null {
  const checks = [
    () => validateNoTeacherConflict(schedule),
    () => validateNoRoomConflict(schedule),
    () => validateTeacherBlockLimit(schedule),
    () => validateNoDuplicateSubjectPerDay(schedule),
    () => validateTeacherSubjects(cfg, schedule),
    () => validateSubjectBlockCounts(cfg, schedule),
    () => validateNoGaps(cfg, schedule),
  ]
  for (const check of checks) {
    const err = check()
    if (err) return err
  }
  return null
}
